### PR TITLE
Language support

### PR DESCRIPTION
**Pull request by jonahsol**
Adds flexibility by allowing users to provide their own parsers for the document and metadata lines. Includes an example of providing a parser for metadata lines in English or Spanish.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -28,6 +30,32 @@ Usage
...
 
+Programmatic Usage
+^^^^^^^^^^^^^^^^^^
+
+.. code:: py
+
+    from clippings import parse_clippings
+
+    my_clippings_fl = ...
+    parse_clippings(my_clippings_fl)
```
Comment (samueldg) [inline README.rst]:
```suggestion
    my_clippings_file = ...
    parse_clippings(my_clippings_file)
```

Worth adding the extra couple characters, for clarity.

File: README.rst [hunk 2]
```diff
@@ -28,6 +30,32 @@ Usage
...
+    - La subrayado en la página 4 | posición 60-60 | Añadido el miércoles, 6 de julio de 2022 06:54:57
+
+    asintiendo a una pregunta que Max no había llegado a formular.
+
+``parse_clippings`` won't parse this by default but you can write your own parser for 
+the second line and pass it to the parameter ``metadata_parser``. 
+You can take a look at an example `here <examples/bilingual_spanish_english_kindle/main.py>`_.
+
+
+
```
Comment (samueldg) [inline README.rst]:
```suggestion
```

Removing extra newlines.

File: README.rst [hunk 3]
```diff
@@ -28,6 +30,32 @@ Usage
...
+^^^^^^^^^^^^^^^^^^
+
+.. code:: py
+
+    from clippings import parse_clippings
+
+    my_clippings_fl = ...
+    parse_clippings(my_clippings_fl)
+
+Want to parse non-English clippings?
```
Comment (samueldg) [inline README.rst:43]: This is awesome! Thanks for adding this to the docs.

By the way, if you want to have a spanish parser into the "main" library instead of just being an example, I'm open to it!

File: clippings/parser.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 import argparse
 import json
 import re
+from typing import Callable
```
Comment (samueldg) [inline clippings/parser.py:5]: Nice! I should eventually add more typing info, especially if we expect more folks to use this library programmatically (as opposed to the CLI).

File: clippings/parser.py [hunk 5]
```diff
@@ -150,7 +151,10 @@ def to_dict(self):
         }
 
 
-def parse_clippings(clippings_file):
+def parse_clippings(
+        clippings_file,
+        document_parser: Callable[[str], Document] = Document.parse,
+        metadata_parser: Callable[[str], Metadata] = Metadata.parse):
```
Comment (samueldg) [inline clippings/parser.py:157]: I like this API. It's backwards compatible, clear and simple. Also using dependency injection like this means we could simplify the tests at some point 👍

File: examples/bilingual_spanish_english_kindle/en_metadata_parser.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+from .utils import *
```
Comment (samueldg) [inline examples/bilingual_spanish_english_kindle/en_metadata_parser.py]: Could we switch this to explicit imports? I know it can be tedious especially for an example file, but `import *` has a few issues that I'd like to avoid:

* Easier to break this example if we move this around
* Harder static analysis
* Possible name collisions

File: examples/bilingual_spanish_english_kindle/en_or_es_metadata_parser.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+from .en_metadata_parser import en_metadata_parser
+from .es_metadata_parser import es_metadata_parser
+import re
+
+
+def en_or_es_metadata_parser(metadata_line: str):
+    """Parse @metadata_line in Spanish or English"""
+    if re.search("Your", metadata_line):
```
Comment (samueldg) [inline examples/bilingual_spanish_english_kindle/en_or_es_metadata_parser.py:8]: Not a big deal especially because this is just an example, but in theory this could have a false positive if that string is included in the title of the book or something. Maybe this could be changed to `re.match` to only check at the beginning of the metadata line?

File: examples/bilingual_spanish_english_kindle/utils.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+from parser import ParserError
+from enum import Enum, auto
+from typing import Optional
+import re
+from datetime import datetime
+import locale as pylocale
```
Comment (samueldg) [inline examples/bilingual_spanish_english_kindle/utils.py]: Why is this rename needed? Does the name clash with something else?

File: examples/bilingual_spanish_english_kindle/utils.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    BOOKMARK = auto()
+
+
+@contextlib.contextmanager
+def setlocale(*args, **kw):
+    """ With thanks: https://stackoverflow.com/a/18594128
+    """
+    saved = pylocale.setlocale(pylocale.LC_ALL)
+    yield pylocale.setlocale(*args, **kw)
+    pylocale.setlocale(pylocale.LC_ALL, saved)
```
Comment (samueldg) [inline examples/bilingual_spanish_english_kindle/utils.py]: Nice fixture!

File: examples/bilingual_spanish_english_kindle/utils.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+def bind_parsers(parser1, parser2):
+    def parser(s: str):
+        res = parser2(s)
+        if res:
+            return parser1(res)
+
+    return parser
+
+
+class MetadataParsers(TypedDict):
```
Comment (samueldg) [inline examples/bilingual_spanish_english_kindle/utils.py:83]: Never used `TypedDict` before... TIL!

**Review by samueldg — COMMENTED** [review]
Awesome stuff! I love that this opens up the lib for tinkering, and possibly many more languages quite easily.

I left a couple comments/questions, but I'd be ok with punting these if you don't have time. Let me know!

I'll also do a release once this merges.

**Comment by jonahsol** [comment]
> Awesome stuff! I love that this opens up the lib for tinkering, and possibly many more languages quite easily.
> 
> I left a couple comments/questions, but I'd be ok with punting these if you don't have time. Let me know!
> 
> I'll also do a release once this merges.

Thanks for the feedback and suggestions! Very much appreciated 😀

File: README.rst (hunk 3 above)
Comment (jonahsol) [reply README.rst:43]: It might not be worth the additional package size and testing requirements, as well as the obligation to include other languages in the same way - Spanish now, French later? But what about Japanese?! I think I'd be more inclined to allow users to do that dependency injection, what do you think?

File: examples/bilingual_spanish_english_kindle/en_metadata_parser.py (hunk 6 above)
Comment (jonahsol) [reply examples/bilingual_spanish_english_kindle/en_metadata_parser.py]: Thanks for the explanation. Resolved by [af698b6](https://github.com/samueldg/clippings/pull/27/commits/af698b6070a95961c869b6efeb22e9f36037c6a7).

File: examples/bilingual_spanish_english_kindle/en_or_es_metadata_parser.py (hunk 7 above)
Comment (jonahsol) [reply examples/bilingual_spanish_english_kindle/en_or_es_metadata_parser.py:8]: Ah that's true! TIL the difference between `re.match` and `re.search` - cheers. I think in this case `re.search` is warranted, since only the metadata line is searched (see `parser.py/parse_clippings()`) and the match string would need to be changed to `"- Your"` for lines of form `"- Your Highlight on page 115 | location 1756-1757 | Added on Friday, 26 February 2021 11:03:02"`.

File: examples/bilingual_spanish_english_kindle/utils.py (hunk 8 above)
Comment (jonahsol) [reply examples/bilingual_spanish_english_kindle/utils.py]: See [0fc6d01](https://github.com/samueldg/clippings/pull/27/commits/0fc6d01e378947a9902ca19316a48e954cb2fc11).

File: README.rst (hunk 3 above)
Comment (samueldg) [reply README.rst:43]: Yeah let's keep the dependency injection for now, and reconsider if there's more demand for language support in general

File: examples/bilingual_spanish_english_kindle/utils.py (hunk 8 above)
Comment (samueldg) [reply examples/bilingual_spanish_english_kindle/utils.py]: Thanks!

**Comment by samueldg** [comment]
@jonahsol Thanks again for this contribution! I created a 0.9.0 release with your new features, plus an assortment of minor changes (see #30) and [uploaded it to PyPI](https://pypi.org/project/clippings/0.9.0/).
